Foundation/NSURLSession/EasyHandle.swift

```diff
@@ -0,0 +1,632 @@
+            var length = Double()
+            try! CFURLSession_easy_getinfo_double(handle.rawHandle, CFURLSessionInfoCONTENT_LENGTH_DOWNLOAD, &length).asError()
+            return handle.didReceive(headerData: data, size: size, nmemb: nmemb, fileLength: length)
+            }.asError()
+        // socket options
+        try! CFURLSession_easy_setopt_ptr(rawHandle, CFURLSessionOptionSOCKOPTDATA, UnsafeMutablePointer<Void>(unsafeAddress(of: self))).asError()
+        try! CFURLSession_easy_setopt_sc(rawHandle, CFURLSessionOptionSOCKOPTFUNCTION) { (userdata: UnsafeMutablePointer<Void>?, fd: CInt, type: CFURLSessionSocketType) -> CInt in
+            guard let handle = URLSessionTask._EasyHandle.from(callbackUserData: userdata) else { return 0 }
+            guard type == CFURLSessionSocketTypeIPCXN else { return 0 }
+            do {
+                try handle.setSocketOptions(for: fd)
+                return 0
+            } catch {
+                return 1
+            }
+            }.asError()
+        // seeking in input stream
+        try! CFURLSession_easy_setopt_ptr(rawHandle, CFURLSessionOptionSEEKDATA, UnsafeMutablePointer<Void>(unsafeAddress(of: self))).asError()
+        try! CFURLSession_easy_setopt_seek(rawHandle, CFURLSessionOptionSEEKFUNCTION, { (userdata, offset, origin) -> Int32 in
+            guard let handle = URLSessionTask._EasyHandle.from(callbackUserData: userdata) else { return CFURLSessionSeekFail }
+            return handle.seekInputStream(offset: offset, origin: origin)
+        }).asError()
+        
+        // progress
+        
+        
+        
+        try! CFURLSession_easy_setopt_long(rawHandle, CFURLSessionOptionNOPROGRESS, 0).asError()
+        
+        try! CFURLSession_easy_setopt_ptr(rawHandle, CFURLSessionOptionPROGRESSDATA, UnsafeMutablePointer<Void>(unsafeAddress(of: self))).asError()
+        
+        try! CFURLSession_easy_setopt_tc(rawHandle, CFURLSessionOptionXFERINFOFUNCTION, { (userdata: UnsafeMutablePointer<Void>?, dltotal :Int64, dlnow: Int64, ultotal: Int64, ulnow: Int64) -> Int32 in
+            guard let handle = URLSessionTask._EasyHandle.from(callbackUserData: userdata) else { return -1 }
+            handle.updateProgressMeter(with: _Progress(totalBytesSent: ulnow, totalBytesExpectedToSend: ultotal, totalBytesReceived: dlnow, totalBytesExpectedToReceive: dltotal))
+            return 0
+        }).asError()
+
+    }
+    /// This callback function gets called by libcurl when it receives body
+    /// data.
+    ///
+    /// - SeeAlso: <https://curl.haxx.se/libcurl/c/CURLOPT_WRITEFUNCTION.html>
+    func didReceive(data: UnsafeMutablePointer<Int8>, size: Int, nmemb: Int) -> Int {
+        URLSession.printDebug("[EasyHandle] -> write callback \(size * nmemb)")
+        let d: Int = {
+            let buffer = UnsafeBufferPointer(start: data, count: size * nmemb)
+            switch delegate.didReceive(data: buffer) {
```

Comment:
As discussed earlier, let's try to get these kinds of things into `struct Data` to present a more sane Swift interface after we've received stuff from the C layer.